boot/freeldr/freeldr/ntldr/registry.c

```diff
@@ -89,11 +111,363 @@ RegImportBinaryHive(
+ * to sync all the dirty data from memory to the backing
+ * storage in order to make the recovery data visible.
+ */
+BOOLEAN
+RegRecoverSystemHive(
+    _Inout_ PVOID ChunkBase,
+    _In_ ULONG_PTR SystemLog)
+{
+    PHBASE_BLOCK SystemHiveBlock, SystemLogBlock;
+    PVOID SystemFlatHive, SystemFlatLog;
+
+    /* Capture the base blocks */
+    SystemHiveBlock = GET_HBASE_BLOCK(ChunkBase);
+    SystemLogBlock = GET_HBASE_BLOCK(SystemLog);
+
+    /*
+     * Validate the base block header of the
+     * SYSTEM hive. We will need to apply a
+     * new header obtained from the log.
+     */
+    if (!HvpVerifyHiveHeader(SystemHiveBlock, HFILE_TYPE_PRIMARY))
+    {
+        /* Recover it */
+        RtlCopyMemory(SystemHiveBlock,
+                      SystemLogBlock,
+                      SystemLogBlock->Cluster * HSECTOR_SIZE);
+        SystemHiveBlock->Type = HFILE_TYPE_PRIMARY;
+    }
+
+    /*
+     * Now copy the data from the SYSTEM log
+     * to the primary hive.
+     */
+    SystemFlatHive = GET_REG_ADDRESS(ChunkBase);
+    SystemFlatLog = GET_REG_ADDRESS(SystemLog);
+    if (!RegCopyDataFromSystemLog(SystemLogBlock,
+                                  SystemHiveBlock->Length,
+                                  SystemFlatHive,
+                                  SystemFlatLog))
+    {
+        return FALSE;
+    }
+
+    /*
+     * Fix up the secondary sequence and set up
+     * a new checksum as we wrote data from the
+     * SYSTEM log into the hive.
+     */
+    SystemHiveBlock->Sequence2 = SystemHiveBlock->Sequence1;
+    SystemHiveBlock->CheckSum = HvpHiveHeaderChecksum(SystemHiveBlock);
+    return TRUE;
+}
+
+/**
+ * @brief
+ * Imports the SYSTEM binary hive from
+ * the registry base chunk that's been
+ * provided by the loader block.
+ *
+ * @param[in] ChunkBase
+ * A pointer to the registry base chunk
+ * that serves for SYSTEM hive initialization.
+ *
+ * @param[in] ChunkSize
+ * The size of the registry base chunk. This
+ * parameter refers to the actual size of
+ * the SYSTEM hive. This parameter is currently
+ * unused.
+ *
+ * @return
+ * Returns TRUE if hive importing and initialization
+ * have succeeded, FALSE otherwise.
+ */
+BOOLEAN
+RegImportBinaryHive(
+    _In_ PVOID ChunkBase,
+    _In_ ULONG ChunkSize)
+{
+    BOOLEAN Success;
+    ULONG_PTR SystemLog;
+    PCM_KEY_NODE KeyNode;
+
+    TRACE("RegImportBinaryHive(%p, 0x%lx)\n", ChunkBase, ChunkSize);
+
+    /* Assume that we don't need boot recover, unless we have to */
+    ((PHBASE_BLOCK)ChunkBase)->BootRecover = HBOOT_NO_BOOT_RECOVER;
+
+    /* Allocate and initialize the hive */
+    CmSystemHive = FrLdrTempAlloc(sizeof(CMHIVE), 'eviH');
+    Success = RegInitializeSystemHive(CmSystemHive, ChunkBase);
+    if (!Success)
+    {
+        /*
+         * The SYSTEM hive is corrupt. Recover
+         * it with the SYSTEM log that we have
+         * at disposal. Free the system hive buffer
+         * first.
```

Comment:
A general note: I won't highlight this everywhere, but when scrolling through your code I do often think that your code would be better readable if you would avoid a few line-breaks here and there within your comments.
Don't get me wrong: Commenting is good, but being too excessive and consuming too many lines with it will make less **actual code** fit to the screen, which makes it harder for readers to grasp the wider context of things.
Please take that thought with you for future PRs.